### ISSUE TITLE
09-k-means-housing.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/josefina-aispuro-merelles-machine-learning#5]
Hola @josefina-aispuro-merelles , por aquí las issues!

## LO QUE HAS HECHO BIEN

- **Eliminación de duplicados**: Correcto con verificación condicional
- **Selección de columnas**: Correctamente eliminas las columnas irrelevantes
- **6 clusters**: Correcto según el ejercicio
- **Modelo supervisado con RandomForest**: Bien implementado
- **Optimización con GridSearchCV**: Excelente, pruebas múltiples hiperparámetros
- **Guardado de modelos**: Perfecto, guardas tanto K-Means como el RandomForest optimizado
- **Estructura clara**: El notebook está bien organizado con secciones claras
- **Conclusiones**: Buenas observaciones sobre el rendimiento

### 1. No pasa nada, pero a tener en cuenta. NO divides train/test ANTES del clustering

El ejercicio pide:
> "Asegúrate de dividir convenientemente el conjunto de datos en train y test... podrás utilizarlos para entrenar el algoritmo no supervisado y luego para hacer predicciones sobre puntos nuevos"

Tú aplicas K-Means sobre **todo el dataset** y luego divides:

```python
# Tu código: K-Means sobre todo df
df["cluster"] = kmeans.fit_predict(df)

# Y DESPUÉS divides
X_train, X_test, y_train, y_test = train_test_split(X, y, ...)
```

**DEBERÍA SER:**
```python
# Primero dividir
X_train, X_test = train_test_split(df, test_size=0.2, random_state=42)

# K-Means solo con train
kmeans = KMeans(n_clusters=6, random_state=42)
kmeans.fit(X_train)

# Predecir en ambos
train_clusters = kmeans.predict(X_train)
test_clusters = kmeans.predict(X_test)
```
---

### 2. Falta escalar los datos (IMPORTANTE)

K-Means es sensible a la escala. `MedInc` tiene un rango diferente a `Latitude` y `Longitude`. Deberías escalar:

```python
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)  # Solo transform, NO fit_transform

kmeans.fit(X_train_scaled)
```
---

### 3. Falta el Paso 3: Predicción con test y visualización comparativa

El ejercicio pide:
> "Ahora utiliza el modelo entrenado con el conjunto test y añade los puntos al gráfico anterior para confirmar que la predicción es satisfactoria"

No muestras ninguna visualización de los clusters ni comparación train vs test.

```python
# Visualización básica de clusters
plt.figure(figsize=(10, 6))
plt.scatter(X_train['Longitude'], X_train['Latitude'], c=train_clusters, cmap='tab10', alpha=0.5, label='Train')
plt.scatter(X_test['Longitude'], X_test['Latitude'], c=test_clusters, cmap='tab10', marker='x', label='Test')
plt.legend()
plt.title('Clusters: Train vs Test')
plt.show()
```

---

### 4. Faltan métricas adicionales del modelo supervisado

Solo muestras accuracy. Para clasificación multiclase deberías incluir:

```python
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred_test))
print(confusion_matrix(y_test, y_pred_test))
```

Tienes `confusion_matrix` importado pero no lo usas.

---

### 5. Falta análisis/interpretación de los clusters

No describes qué representa cada cluster. Deberías añadir algo como:

```python
# Perfil de cada cluster
df.groupby("cluster")[["MedInc", "Latitude", "Longitude"]].mean()
```

Y una interpretación: "Cluster 0 = zonas de ingresos bajos en el norte", etc.

---

El notebook tiene una estructura muy limpia y la parte de optimización con GridSearchCV está excelente. Los puntos principales a corregir son: seguir el flujo del ejercicio (split → escalar → K-Means en train → predict en test), añadir visualizaciones de los clusters, y hacer un análisis interpretativo de qué significa cada cluster geográfica y económicamente.
